### ISSUE TITLE
Number of trainable parameters

### ISSUE THREAD

**Issue by tmanh** [JaouadT/KANU_Net#1]
Well,
I think it is nice, however, I would like to point out that your experiments do not have the same experimental setup.

Total trainable parameters  for your KANU_Net is 33470626.
Total trainable parameters  for your U_Net is only 17262977. So, KANU_Net is nearly double the size of the UNet, this is not a fair comparison.

When you used KAN, you add more trainable parameters to the network, which is not fair for the original UNet anymore. So, the question, for example, should be adding KAN layers or adding more convolutional layers, which one is more effective? You should do experiments to examine this instead of compare to the original architecture.

I have done multiple test from classification, super-resolution, segmentation, etc. For easy datasets, adding KAN layers can improve the performance of the networks. However, for more complicated datasets, what I observed is that adding KAN layers is actually reduce the performance of the original architecture. So, I guess you should also test it on multiple datasets before drawing the conclusion.

**Comment by JaouadT** [comment]
Hello,
First of all what do you mean by "experiments do not have the same experimental setup"?, should we have different architecture with the same number of parameters to compare them? is that what you mean by same experiments setup?

I think that the comparison makes sense, since when you use the KAN convolution, you add extra parameters as you stated. So, I kept the original U-Net architecture untouched when it comes to the other components, except for the convolution layers so that I can have a U-Net based on KAN convolution. Generally, when you compare architectures, you perform your experiments with the same training and validating setup, same dataset, etc. but the models differs. Now in terms of models complexity that's another analysis. Besides I didn't draw any conclusions I only presented my findings when testing both architectures on the dataset.

**Comment by tmanh** [comment]
No, your comparison indeed does not make sense. You should not just compare the original Unet with Unet+KAN. UNet + KAN have nearly 2 times more parameters compared to the original UNet. How can this is a fair comparison between two architectures? Furthermore, with your current setup, the KANConv has two convolutional layers, so you already made the numbers of convolutional layers double? It’s not the same architecture anymore.

And, it is not about model complexity. You can increase the accuracy of the model by adding more layers to your current one. Now, you added so many parameters to the point the new model is x2 bigger than the old one. Of course, without testing the model, I would say it often is better. If you just compared like that, you do not really need to do the comparison actually.

To have the same setup, you should have roughly the same size network, and seeing which way is better to improve the accuracy. If adding few more convolutional layers to the UNet can make it is better than KAN. Do you really need KAN?

**Comment by JaouadT** [comment]
Take for example the work TransUnet, in their paper they compared their method against methods like V-Net, ViT, etc. according to you surely these methods don't have the same number of parameters so they shouldn't be compared. 
<img width="1323" alt="Screenshot 2024-05-24 at 16 40 17" src="https://github.com/JaouadT/KANU_Net/assets/75540016/7f3c513f-9914-4eed-9e31-f58748cd424a">

For your question, I don't know if you can change a network's layers or the components of it and except to have same number of parameters.

**Comment by tmanh** [comment]
First, there are two types of comparison. The one you showed is one of them. You either add something to the network or replace something with something else. So, you, of course, need to compare the original model with the new one to ensure that your model is actually better than the old one. Yes, there are many papers that since they focus more on the applications.

However, in your case, the idea of KAN is not just a new activation function; it is equivalent to MLP + activation function. If you read the original paper on KAN, they do not compare a KAN with more parameters than an MLP + activation function. Instead, they compare a KAN with fewer parameters than an MLP + activation function to show that KAN is more efficient. This is a more important comparison since it shows which model is more efficient rather than just focusing on improving accuracy by increasing the complexity of the network. This is why people are excited about KAN. People want to know if this idea can be extended to more complex problems or not.

For the ConvKAN where you copied the code, they used two Conv2D layers instead of one. Thus, even though in your code, the Unet and KANUnet are identical except one uses nn.Conv2D and the other uses FastKANConvLayer, the difference is that the self.double_conv layer in KANUnet has four nn.Conv2D inside it while the Unet only has two Conv2D. So, you did not show the efficiency of KAN here since KANUnet is much bigger than the Unet you implemented. I would say these experiments are not that interesting.

**Comment by JaouadT** [comment]
We agree on this: 
You either add something to the network or replace something with something else. So, you, of course, need to compare the original model with the new one to ensure that your model is actually better than the old one. Yes, there are many papers that since they focus more on the applications.

**Comment by JackHeroFly** [comment]
Hello, thank you for sharing the code. May I ask how you divided the dataset? The code does not include steps on how to divide the BUSI dataset, but directly loads the divided data. Can you provide this part of the code?
